cpu/atmega_common/include/gpio_ll_arch.h

```diff
@@ -0,0 +1,271 @@
+#define GPIO_LL_ARCH_H
+
+#include <assert.h>
+
+#include "cpu.h"
+#include "irq.h"
+#include "kernel_defines.h"
+#include "periph_cpu.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef DOXYGEN /* hide implementation specific details from Doxygen */
+
+#ifdef PORTH
+#define GPIO_PORT(num) \
+    ((num >= PORT_H) ? \
+     (ATMEGA_GPIO_BASE_H + ((num) - PORT_H) * ATMEGA_GPIO_SIZE) : \
+     (ATMEGA_GPIO_BASE_A + (num) * ATMEGA_GPIO_SIZE))
+#define GPIO_PORT_NUM(port) \
+    (((port) >= ATMEGA_GPIO_BASE_H) ? \
+     (((port) - ATMEGA_GPIO_BASE_H) / ATMEGA_GPIO_SIZE + PORT_H) : \
+     (((port) - ATMEGA_GPIO_BASE_A) / ATMEGA_GPIO_SIZE))
+#else
+#define GPIO_PORT(num) (ATMEGA_GPIO_BASE_A + (num) * ATMEGA_GPIO_SIZE)
+#define GPIO_PORT_NUM(port) (((port) - ATMEGA_GPIO_BASE_A) / ATMEGA_GPIO_SIZE)
+#endif
+
+static inline uword_t gpio_ll_read(gpio_port_t port)
+{
+    atmega_gpio_port_t *p = (void *)port;
+    return p->pin;
+}
+
+static inline uword_t gpio_ll_read_output(gpio_port_t port)
+{
+    atmega_gpio_port_t *p = (void *)port;
+    return p->port;
+}
+
+/**
+ * @brief   Check if a set/clear operation with the on the given port with the
+ *          given mask can be done in a single-cycle bit-set / bit-clear
+ *          instruction.
+ *
+ * @param   port    GPIO port to set/clear bits on
+ * @param   mask    Bitmask specifying the bits to set/clear
+ *
+ * The compiler can optimize a clear/set operation if all of the following is
+ * true:
+ *
+ * 1. The GPIO port number is known at compile time, so that it can be used as
+ *    immediate
+ * 2. The GPIO port is alphabetically smaller than H, otherwise its registers
+ *    are not within the I/O space (and cannot fit the 5 bit immediate field)
+ * 3. The bitmask is known at compile time
+ * 4. Exactly one bit is set in the bitmask
+ */
+static inline bool _can_bitwise_access(gpio_port_t port, uword_t mask)
+{
+    if (IS_CT_CONSTANT(port)
+#ifdef ATMEGA_GPIO_BASE_H
+        && IS_CT_CONSTANT(port < ATMEGA_GPIO_BASE_H)
+        && (port < ATMEGA_GPIO_BASE_H)
+#endif
+        && IS_CT_CONSTANT(mask)
+        && IS_CT_CONSTANT(__builtin_popcount(mask) == 1)) {
+        return __builtin_popcount(mask) == 1;
+    }
+
+    return 0;
+}
+
+static inline void gpio_ll_set(gpio_port_t port, uword_t mask)
+{
+    atmega_gpio_port_t *p = (void *)port;
+    if (_can_bitwise_access(port, mask)) {
+            p->port |= mask;
+    }
+    else {
+        unsigned state = irq_disable();
+        p->port |= mask;
+        irq_restore(state);
+    }
+}
+
+static inline void gpio_ll_clear(gpio_port_t port, uword_t mask)
+{
+    atmega_gpio_port_t *p = (void *)port;
+    if (_can_bitwise_access(port, mask)) {
+        p->port &= ~mask;
+    }
+    else {
+        unsigned state = irq_disable();
+        p->port &= ~mask;
+        irq_restore(state);
+    }
+}
+
+static inline void gpio_ll_toggle(gpio_port_t port, uword_t mask)
+{
+    atmega_gpio_port_t *p = (void *)port;
+    p->pin = mask;
```

Comment:
Not
```suggestion
    p->pin ^= mask;
```
?

Reply:
No, the ATmegas have three registers per GPIO, the data direction register (DDR), the PIN and the port register. Those have multiple overloaded use, as only the first 64 addresses can be used for fast access.

The DDR register is used to configure the direction. The PORT register is the output register, but for input pins it configured the pull up resistor instead. The PIN register is the input register, but writes to it toggle bits in the output register instead.

I don't know why they have used better names (well, the DDR name is fine) for the registers to better reflect their use.

Reply:
Better add a comment then to avoid future confusion.

Reply:
I added a comment, as this is indeed very non-obvious.